format/wasm/instructions.go

```diff
@@ -0,0 +1,1539 @@
+	case 0xc3:
+		decodeInstructionWithoutOperands(d, "i64.extend16_s", i)
+	case 0xc4:
+		decodeInstructionWithoutOperands(d, "i64.extend32_s", i)
+
+	case 0xd0:
+		decodeRefNull(d)
+	case 0xd1:
+		decodeInstructionWithoutOperands(d, "ref.is_null", i)
+	case 0xd2:
+		decodeRefFunc(d)
+
+	case 0xfc:
+		decodePrefixedInstruction(d)
+
+	case 0xfd:
+		decodeVectorInstruction(d)
+	default:
+		d.Fatalf("unknown instruction: %#02x", instr[0])
+	}
+}
+
+func decodeInstructionWithoutOperands(d *decode.D, name string, i byte) {
+	d.FieldU8(name, d.AssertU(uint64(i)), scalar.ActualHex)
+}
+
+func decodeBlock(d *decode.D) {
+	d.FieldU8("block", d.AssertU(0x02), scalar.ActualHex)
+	decodeBlockType(d, "bt")
+	d.FieldArray("instructions", func(d *decode.D) {
+		for {
+			b := d.PeekBytes(1)[0]
+			if b == 0x0b {
+				break
+			}
+			d.FieldStruct("instr", decodeInstruction)
+		}
+	})
+	d.FieldU8("end", d.AssertU(0x0b))
+}
+
+func decodeLoop(d *decode.D) {
+	d.FieldU8("loop", d.AssertU(0x03), scalar.ActualHex)
+	decodeBlockType(d, "bt")
+	d.FieldArray("instructions", func(d *decode.D) {
+		for {
+			b := d.PeekBytes(1)[0]
+			if b == 0x0b {
+				break
+			}
+			d.FieldStruct("instr", decodeInstruction)
+		}
+	})
+	d.FieldU8("end", d.AssertU(0x0b))
+}
+
+func decodeIf(d *decode.D) {
+	d.FieldU8("if", d.AssertU(0x04), scalar.ActualHex)
+	decodeBlockType(d, "bt")
+	elseClause := false
+	d.FieldArray("in1", func(d *decode.D) {
+		for {
+			b := d.PeekBytes(1)[0]
+			if b == 0x05 {
+				elseClause = true
+				break
+			}
+			if b == 0x0b {
+				break
+			}
+			d.FieldStruct("instr", decodeInstruction)
+		}
+	})
+	if elseClause {
+		d.FieldU8("else", d.AssertU(0x05), scalar.ActualHex)
+		d.FieldArray("in2", func(d *decode.D) {
+			for {
+				b := d.PeekBytes(1)[0]
+				if b == 0x0b {
+					break
+				}
+				d.FieldStruct("instr", decodeInstruction)
+			}
+		})
+	}
+	d.FieldU8("end", d.AssertU(0x0b))
+}
+
+func decodeEnd(d *decode.D) {
+	d.FieldU8("end", d.AssertU(0x0b), scalar.ActualHex)
+}
+
+func decodeBr(d *decode.D) {
+	d.FieldU8("br", d.AssertU(0x0c), scalar.ActualHex)
+	decodeLabelIdx(d, "l")
+}
+
+func decodeBrIf(d *decode.D) {
+	d.FieldU8("br_if", d.AssertU(0x0d), scalar.ActualHex)
+	decodeLabelIdx(d, "l")
+}
+
+func decodeBrTable(d *decode.D) {
+	d.FieldU8("br_table", d.AssertU(0x0e), scalar.ActualHex)
+	decodeVec(d, "l", func(d *decode.D) {
+		decodeLabelIdx(d, "l")
+	})
+	decodeLabelIdx(d, "lN")
+}
+
+func decodeCall(d *decode.D) {
+	d.FieldU8("call", d.AssertU(0x10), scalar.ActualHex)
+	decodeFuncIdx(d, "x")
+}
+
+func decodeCallIndirect(d *decode.D) {
+	d.FieldU8("call", d.AssertU(0x11), scalar.ActualHex)
+	decodeTypeIdx(d, "y")
+	decodeTableIdx(d, "x")
+}
+
+func decodeSelectT(d *decode.D) {
+	d.FieldU8("select", d.AssertU(0x1c), scalar.ActualHex)
+	decodeVec(d, "t", func(d *decode.D) {
+		decodeValType(d, "t")
+	})
+}
+
+func decodeLocalGet(d *decode.D) {
+	d.FieldU8("local.get", d.AssertU(0x20), scalar.ActualHex)
```

Comment:
Probably a good idea to only use valid jq identifiers as field name, ex: `.something.local_get` vs `.something."local.get"` or `.something.["local.get"]`

Reply:
I'll change the field name to `opcode` or something, and have the mnemonic as symbol value.

Reply:
When writing decoders i usually try to think how to make the structure easy to work with for doing queries etc. For example maybe group things into structs that might be useful as one unit and also try to avoid repetition that would make queries long to write.